subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/DependencyGraphEdge.java

```diff
@@ -41,9 +40,9 @@
     Iterable<? extends DependencyGraphNode> getTargets();
 
     /**
-     * The original dependency instance declared in the build script, if any.
+     * The original dependency or dependency constraint instance declared in the build script, if any.
      */
     @Nullable
-    Dependency getOriginalDependency();
+    Object getOriginalDependencyOrConstraint();
```

Comment:
:x: The way Adam describes it, a constraint does not result in any edge in the graph. In that case, there should always be an 'original dependency' for every edge. I'm not sure what this means for the way we process pending dependencies: perhaps we can do less work by not adding them as separate edges by instead just using them to influence version selection. It feels like some major rework may be involved.